test/box-luatest/gh_10583_support_async_wait_modes_for_sync_tx_test.lua

```diff
@@ -0,0 +1,206 @@
+local server = require('luatest.server')
+local t = require('luatest')
+
+local test_matrix = {
+    wait_mode = {'submit', 'none'},
+    engine = {'memtx', 'vinyl'},
+    memtx_use_mvcc_engine = {false, true},
+}
+local g = t.group('common', t.helpers.matrix(test_matrix))
+table.insert(test_matrix.wait_mode, 'complete')
+local g_corner_cases = t.group('corner_cases', t.helpers.matrix(test_matrix))
+
+local function setup_server(cg)
+    cg.s = server:new({
+        alias = 'default',
+        box_cfg = {
+            memtx_use_mvcc_engine = cg.params.memtx_use_mvcc_engine,
+            replication_synchro_timeout = 120,
+            replication_synchro_quorum = 2,
+        },
+    })
+    cg.s:start()
+    cg.s:exec(function(engine)
+        rawset(_G, 'fiber', require('fiber'))
+        rawset(_G, 'are_txns_not_isolated',
+               not box.cfg.memtx_use_mvcc_engine and engine ~= 'vinyl')
+
+        box.schema.create_space('a', {engine = engine,
+                                      is_sync = false}):create_index('pk')
+        box.schema.create_space('s', {engine = engine,
+                                      is_sync = true}):create_index('pk')
+        box.ctl.promote()
+    end, {cg.params.engine})
+end
+
+g.before_each(setup_server)
+
+g.after_each(function(cg)
+    cg.s:drop()
+end)
+
+g.test_wait_modes_for_sync_tx = function(cg)
+    cg.s:exec(function(wait_mode)
+        box.atomic({wait = wait_mode}, function() box.space.s:replace{0} end)
+        t.assert(box.info.synchro.queue.len, 1)
+
+        box.begin{txn_isolation = 'read-committed'}
+        t.assert_equals(box.space.s:get{0}, {0})
+        box.commit()
+        box.begin{txn_isolation = 'read-confirmed'}
+        t.assert_equals(box.space.s:get{0},
+                        _G.are_txns_not_isolated and {0} or nil)
+        box.commit()
```

Comment:
Before and after this function the counter `fiber.self():csw()` must remain unchanged when `wait_mode='none'`.

Reply:
Added checks for this.